### ISSUE TITLE
Fix "error: [Errno 98] Address already in use" exposed during testing

### ISSUE THREAD

**Issue by breerly** [uber/tchannel-python#256]
Tests have become rather flaky, due to the issue `error: [Errno 98] Address already in use` which can be viewed in the following test:

https://travis-ci.org/uber/tchannel-python/jobs/85838961

```
tests/tornado/test_connection.py ..
tests/tornado/test_dispatch.py ....
tests/tornado/test_hyperbahn.py .....
tests/tornado/test_peer.py ...........
tests/tornado/test_request.py .
tests/tornado/test_tchannel.py .....
tests/tornado/test_timeout.py .

==================================== ERRORS ====================================
__________ ERROR at setup of test_set[client_thriftrw-server_thrift] ___________

io_loop = <tornado.platform.epoll.EPollIOLoop object at 0x2aae9b7dd110>

[1m    @pytest.fixture[0m
[1m    def server(io_loop):  # need io_loop fixture for listen() to work[0m
[1m        server = TChannel(name='server')[0m
[1m>       server.listen()[0m

tests/schemes/test_thrift.py:65: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tchannel/tchannel.py:227: in listen
[1m    return self._dep_tchannel.listen(port)[0m
tchannel/tornado/tchannel.py:259: in listen
[1m    sockets = bind_sockets(self._port)[0m
.tox/py27/lib/python2.7/site-packages/tornado/netutil.py:187: in bind_sockets
[1m    sock.bind(sockaddr)[0m
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

name = 'bind', self = <socket._socketobject object at 0x2aae9b7dffa0>
args = (('0.0.0.0', 37892),)

[1m    def meth(name,self,*args):[0m
[1m>       return getattr(self._sock,name)(*args)[0m
[1m[31mE       error: [Errno 98] Address already in use[0m

/opt/python/2.7.9/lib/python2.7/socket.py:224: error
[1m[31m=============== 430 passed, 11 skipped, 1 error in 9.80 seconds ================[0m
[31mERROR: InvocationError: '/home/travis/build/uber/tchannel-python/.tox/py27/bin/py.test --cov-report=term-missing tests'[0m
___________________________________ summary ____________________________________
[31mERROR:   py27: commands failed[0m
make: *** [test_ci] Error 1
travis_time:end:11f2f308:start=1445036649983119834,finish=1445036718931727174,duration=68948607340
[0K
[31;1mThe command "make test_ci" exited with 2.[0m

Done. Your build exited with 1.
```  

This is a costly bug - let's fix it once and for all.

**Comment by breerly** [comment]
Possible solution - add a retry when this error is raised. This isn't the prettiest solution, as it means we're punting on understanding what exactly is happening, but I predict adding a retry (up to 3?) in `TChannel.list` when a `socker.error` is raised, and the `errorno` attrib is `98`:

```python
try:
    bindsockets(...)
catch socket.error as e:
    if e.errorno == 98:    # Address already in use
       # retry up to 3-5 times
````

**Comment by abhinav** [comment]
We vendored `bind_sockets` and strategically placed some print statements. As per https://travis-ci.org/uber/tchannel-python/jobs/86301434, you'll note that `bind_sockets` gets called once and `sock.bind` (and incidentally `sock.listen`) twice: once with the IPv6 addr and once with the v4. The v6 one gets the OS assigned port and the v4 one then fails to bind with "Address already in use".

I'm experimenting with passing `family=socket.AF_INET` to `bind_sockets` to make it use IPv4 only for now.

**Comment by breerly** [comment]
Created a PR to 0.18 ^ if the ipv4-only workaround is real.

**Comment by breerly** [comment]
This work-around is looking good, and seeing how we don't officially support IPv6 yet, I'm ok with going forward with it in the morning. Let's keep restarting these jobs to see if we can make it happen again.

**Comment by breerly** [comment]
Resolved by #257
